core/domain/classifier/base_classification_algorithm.py

```diff
@@ -0,0 +1,92 @@
+# coding: utf-8
+#
+# Copyright 2015 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Base class of classification algorithms for free-form text answers"""
+
+from abc import abstractmethod
+
+
+class BaseClassificationAlgorithm(object):
+    """A base class for classifiers that uses supervised learning to match
+    free-form text answers to answer groups. The classifier trains on answers
+    that exploration editors have assigned to an answer group. Given a new
+    answer, it predicts the answer group.
+
+    Below are some concepts used in this class.
+    training_data: list of 'training_data'. Each training_data is represented
+        by a 2-element list. The first item of the list is the single training
+        data, and the second item is a list of labels that the data should
+        be matched.
+    predicting_data: list of 'predicting_data'. Each element of the list
+        represents a single 'predicting_data'.
+    label - An answer group that the doc should correspond to. If a doc is
+        being added to train a model, labels are provided. If a doc is being
+        added for prediction purposes, no labels are provided. If a doc does
+        not match any label, the doc should have only one label, '_default'.
+
+    Attributes:
+        DEFAULT_LABEL: str. The label used to characterize a word with no label
+            assigned to it.
+    """
+
+    DEFAULT_LABEL = '_default'
+
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def from_dict(self, model):
+        """Initializes the properties of this classifier from a dict
+          constructed using to_dict().
+
+          Args:
+              model: A dict representing the classifier.
+        """
+        pass
```

Comment:
Should these be `raise NotImplementedError` instead? We do that in e.g. https://github.com/oppia/oppia/blob/develop/core/tests/test_utils.py#L145, so I was wondering, what are the pros/cons of abc.abstractmethod over this?

(I ask because we probably want to stick to one pattern for this codebase-wide, so if the latter approach is better we might want to do a global refactor. /cc @BenHenning)